### ISSUE TITLE
LG-50 mpMetrics-2.0 implementation

### ISSUE THREAD

**Issue by fmhwong** [OpenLiberty/open-liberty#7946]
AHA: https://bigblue.aha.io/features/LG-50
Original Epic in WS-CD: https://github.ibm.com/was-liberty/WS-CD-Open/issues/11720
 ID for mpMetrics-2.0 https://github.com/OpenLiberty/docs/issues/332

**Comment by fmhwong** [comment]
WAD approval in https://bigblue.aha.io/features/LG-15

**Comment by gscottj** [comment]
This feature does not add any new or change any existing UI.  No AVT is required. Accessibility focal approved.

**Comment by fmhwong** [comment]
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## List of Steps to complete or get approvals / sign-offs for Onboarding to the Liberty release (GM date)
 
Instructions:
 - Do the actions below and mark them complete in the checklist when they are done.
 - Make sure all feature readiness approvers put the appropriate tag on the epic to indicate their approval.
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE **Before Development Starts or 8 weeks before Onboarding** 
- [x] POC Design / WAD Review Scheduled (David Chang) or N/A. AHA Feature marked.
- [x] POC Design / WAD Reviewed (Feature Owner) or N/A. AHA Feature marked.
- [x] Complete any To-Do's from the POC Review, and create a new TO-DO in the AHA Feature for the "Design Approval Request" and assign it to Ian Robinson. Leave the date blank. (Feature Owner)
- [x] Design / WAD Approval (Ian Robinson) or N/A. AHA Feature marked.
- [x] No Design / No WAD Approval (Arthur De Magalhaes - cloud / Alasdair Nottingham - server) or N/A. AHA Feature marked.
- [x] AHA Feature has link to Github epic or issue. Github epic has link back to AHA Feature. (Feature Owner or Backlog Subtribe PM).
- [x] SVT Requirements identified. (Epic owner / Feature owner with SVT focal point)
- [x] ID Requirements identified. (Epic owner / Feature owner with ID focal point)
- [x] Create a child task of the epic entitled "FAT Approval Test Summary"
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE **3 weeks before Onboarding** 
- [x] Identify all open source libraries that are changing or are new. Work with Legal Release Services (Cass Tucker or Release PM) to get open source cleared and approved. Or N/A. (Epic Owner).   New or changed open source impacts license and Certificate of Originality.
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE ** 3 weeks before Onboarding** 
- [ ] All new or changed PII messages are checked into the integration branch, before the last translation shipment out. (Epic Owner)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## TARGET COMPLETION DATE **2 weeks before Onboarding** 
- [x] Implementation complete. (Epic owner / Feature owner) 
- [ ] All function tests complete. Ready for FAT Approval. (Epic owner / Feature owner)
- [ ] Review all known issues for Stop Ship. (Epic owner / Feature owner / PM)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## APPROVALS with TARGET COMPLETION DATE **2 to 1 week before Onboarding** 
Prereq: You must have the Design Approval or No Design Approval on the AHA Feature.
- [x] **Accessibility** - (G Scott Johnston). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [ ] **FAT Liberty SOE** - (Kevin Smith). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [ ] **Globalization** (Marika Joannidis - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.
- [ ] **ID** - (Kareen Deen). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
- [ ] **Performance** - (Jared Anderson). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [ ] **Serviceability** - (Don Bourne). Serviceability has been addressed. 
- [ ] **STE** - (Swati Kasundra). STE chart deck is complete or N/A . Approver adds label focalApproved:ste to the Epic in Github.
- [ ] **SVT** - (Greg Ecock - Cloud, Brian Hanczaryk- APS). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github.
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## TARGET COMPLETION DATE **1 week before Onboarding** 
- [ ] No Stop Ship issues for the feature. (Epic owner / Feature owner / Release PM)
- [ ] Ship Readiness Review and Release Notes completed (Epic owner / Feature owner / Release PM)
- [ ] Github Epic and Epic's issues are closed / complete. All PRs are committed to the master branch. (Epic owner / Feature owner / Backlog Subtribe PM)
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 ## NOT REQUIRED FOR A FEATURE
- [ ] **OL Guides** - (Yee-Kang Chang). Assessment for OL Guides is complete or N/A. 
- [ ] **WDT** - (Leonard Theivendra). WDT work complete or N/A.
 
 
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## Related Deliverables TARGET COMPLETION DATE **General Availability** 
- [ ] Blog article writeup (Epic owner / Feature owner / Laura Cowen)

**Comment by fmhwong** [comment]
@jhanders34 There is no significant change in performance.  May I get your approval on performance?

**Comment by donbourne** [comment]
**Serviceability Approval Comment** - Please answer the following questions for serviceability approval:

1.  WAD -- does the WAD identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation?

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).  
a) What problem paths were tested and demonstrated? 
b) Who did you demo to?
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?  

4.  Which L2 / L3 queues will handle PMRs for this feature?  Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info.

**Comment by marikaj123** [comment]
@fmhwong - Please ensure not to GA the feature before receiving all focal point approvals.

**Comment by skasund** [comment]
L2 has requested STE slides for this feature.  The STE template can be found at the link below: 

https://ibm.box.com/s/1an42g7zdgmaj84w7dft0indqfgi8ffm

Please upload the completed slides to the same STE Archive BOX folder.  Thanks!

**Comment by dave-waddling** [comment]
**FAT Focal Approval comment:** Hi, I've reviewed the Feature Test Summary and discussed the existing test bug defects with @Channyboy in #7949. I'm going to grant the FAT Focal Approval for this feature given that I've made the defects high impact to ensure they are addressed as a priority.

**Comment by jhanders34** [comment]
The performance team put in a change with PR #8080 in order to reduce the performance difference between 2.0 and 1.x. There is still about a 5% performance difference between 2.0 and 1.x for Metrics.  Based off of where the difference being related to new function in the spec and it being only 5%, the performance team deems it to be acceptable and are giving focal approval.  We will track this difference in the future to see if anything can be done to improve it.  Future work is being tracked with https://github.com/OpenLiberty/open-liberty/issues/8078.

**Comment by Channyboy** [comment]
@donbourne 
Serviceability Approval Comment - Please answer the following questions for serviceability approval:

1. WAD -- yes

2. Test and Demo -- MicroProfile Metrics 2.x is an enhanced version of MicroProfile metrics 1.x
It encounters the same serviceability problems as MicroProfile Metrics 2.x. It does not require a new test and demo.

3. SVT 
a) Tanya Kulik
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?
@tkulik  Can you please verify this question?

4. LL3ANA,103

**Comment by tkulik** [comment]
For 3.b:

Yes, I agree that the serviceability is sufficient.

**Comment by skasund** [comment]
@Channyboy  I'm providing the STE approval early.  Continue working on the slides and provide them as soon as possible.

**Comment by Channyboy** [comment]
ID item is: https://github.com/OpenLiberty/docs/issues/332
@chirp1 May you approve this feature.

**Comment by Channyboy** [comment]
See : https://github.com/OpenLiberty/docs/issues/332
As ID item to track.

**Comment by dave-waddling** [comment]
**FAT Focal Approval comment:** Hi, @fmhwong and @Channyboy, I discussed this feature's FAT Focal Approval with one of the Development Leads (@tevans78) and the Test Architect this morning, initially regarding the "inherited" test defects listed in the Feature Test Summary. 

We also took a closer look at the defects, in particular 261878, as this was claimed to "look like" a test bug [here](https://github.com/openliberty/open-liberty/issues/7949#issuecomment-507718122). It's our opinion that this is actually a product bug so please can you:
- [ ]  Investigate the root cause of 261878;
- [ ] Make a definitive statement on the Feature Test Summary as to whether this a test bug or product bug, giving your reasoning.

In the meantime I have withdrawn the FAT Focal Approval.

**Comment by Channyboy** [comment]
For 261878 - (also will be written in the Test Summary)
To summarize, that specific test is missing a vendor metric. This also belongs to the mpMetrics-2.0-monitor-1.0 auto-feature. The root cause is that an mbean  is not able to be retrieved from the mbean server of the JVM. Manual tests shows that starting up a server with the mpMetrics and monitor (which will start up the mpMetrics-2.0-monitor-1.0 autofeature) will get your the `threadpool_default_executor_active_threads`. This leads me to believe its a _timing_ test issue or the an issue with the JVM

**Comment by dave-waddling** [comment]
**FAT Focal Approval comment:** Hi @Channyboy, thanks for the update.

Regarding 261878: More investigation is needed as I don't think it can be a JVM issue - it's occurred on Java versions 8, 11, and 12. It won't be possible to approve this feature until we have a solid handle on the issue, especially as it has occurred intermittently in our continuous builds.

Regarding 260682: Our Windows build engines are in the EBC like all of our other build engines, thus have the same hardware backing them, thus they are equally performant. Also this defect is actually only occurring on the combination of Windows and Java 8 (both IBM and Oracle). Again, more investigation is needed before a decision can be made.

Regarding 263198 and 262549: Agreed, these are known issues with file locking on Windows, these shouldn't prevent getting the FAT Focal Approval.

**Comment by Channyboy** [comment]
261878: (querying mbean server) -
So the metrics-monitor code is (when starting up)  registering the vendor metrics by querying the mbean server. If we get nothing there, we can't create. That is currently the extend of how far we can investigate at the moment.

260682:  (slowness with Windows)
I'ts a known issue (been discussed before with RedHat since they've had similar issues with their windows platforms). I think a variety of things have been modified to the TCK to make it more _leniant_. This can be an issue brought up once again in the Metrics hangouts. But this would not be a product issue.

**Comment by Channyboy** [comment]
@skasund  LG-50 STE doc uploaded

**Comment by skasund** [comment]
@Channyboy   Thanks for the slides!

**Comment by marikaj123** [comment]
@fmhwong - Was the feature code enabled and included into the Liberty 19007 GM build?

**Comment by tevans78** [comment]
Yes, it was all delivered. Only thing missing from this epic is the FAT focal approval.

**Comment by marikaj123** [comment]
Keeping Target:ga as a reminder us that FAT is required.

**Comment by sabolo** [comment]
Please can you provide an update on the progress of the defects holding up this feature. Once you believe you have fixed in a release build for them all we can check the following SOE and ensure that the issues have gone away. Many thanks

**Comment by dave-waddling** [comment]
**FAT Focal Approval comment:** @Channyboy and I had a brief discussion on Slack about the outstanding defects associated with this feature. 

The only product issue that was outstanding (261878) has been resolved and I've confirmed this is the case in the latest SOE. 

The remaining defects are either due to file-locking issues on Windows (262549 and 263198) or a tight timing window in the TCK (260682) which is also being seen by non-IBM companies.

Given the file-locking issues are well understood and that the TCK timing issue is currently outside of our control I'm willing to grant approval at this time.

**Comment by fmhwong** [comment]
Got all approvals and feature shipped in 19.0.0.7
